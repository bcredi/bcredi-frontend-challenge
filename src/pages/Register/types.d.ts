type User =  {
  email: string;
  cpf: string;
  birthDate: string;
  password: string;
}

type InputError = {
  emailError: boolean;
  cpfError: boolean;
  birthDateError: boolean;
  passwordError: boolean;
  readCheckError: boolean;
}

export {User, InputError}