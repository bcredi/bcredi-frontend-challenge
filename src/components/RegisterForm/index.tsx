import React, { FormEvent, ReactNode } from "react";
import styled from "styled-components";
import { User, InputError } from "../../pages/Register/types";
import ActionButton from "../ActionButton";

interface props {
  formTitle: string;
  formDescription: string;
  handleSubmit: () => void;
  children?: ReactNode;
  footer?: ReactNode;
  setData: React.Dispatch<React.SetStateAction<User>>;
  data: User;
  setFormErrors: React.Dispatch<React.SetStateAction<InputError>>;
  formErrors: InputError;
  readCheck: boolean;
}

const RegisterFormContainer = ({
  formTitle,
  formDescription,
  handleSubmit,
  children,
  footer,
  readCheck,
  formErrors,
  setFormErrors,
  setData,
  data,
}: props) => {
  function resetForm() {
    setData({ email: "", cpf: "", birthDate: "", password: "" });
    setFormErrors({
      birthDateError: false,
      cpfError: false,
      emailError: false,
      passwordError: false,
      readCheckError: false,
    });
  }

  const onSubmit = (e: FormEvent) => {
    e.preventDefault();
    if (!data.email || data.email === "") {
      setFormErrors({ ...formErrors, emailError: true });
    } else if (!data.cpf || data.cpf === "") {
      setFormErrors({ ...formErrors, cpfError: true });
    } else if (!data.birthDate || data.birthDate === "") {
      setFormErrors({ ...formErrors, birthDateError: true });
    } else if (!data.password || data.password === "") {
      setFormErrors({ ...formErrors, passwordError: true });
    } else if (!readCheck) {
      setFormErrors({ ...formErrors, readCheckError: true });
    } else {
      handleSubmit();
      resetForm();
    }
  };

  return (
    <RegisterFormWrapper>
      <FormTitle className="form-title">{formTitle}</FormTitle>
      <FormText className="form-description">{formDescription}</FormText>
      <Form name="register" onSubmit={onSubmit}>
        {children}
        <ActionButton text="Cadastrar" type="submit" />
      </Form>
      <FormFooter className="form-footer">{footer}</FormFooter>
    </RegisterFormWrapper>
  );
};

const FormTitle = styled.div`
  width: 312px;
  height: 30px;
  font-weight: bold;
  font-size: 24px;
  line-height: 30px;
  color: var(--base-color-text);

  @media (min-width: 600px) {
    width: 320px;
    text-align: center;
  }
`;

const FormText = styled.div`
  height: 54px;
  font-size: 16px;
  line-height: 24px;
  color: var(--base-color-warm-grey);
  margin: 6px auto 0px auto;

  @media (min-width: 600px) {
    text-align: center;
  }
`;

const FormFooter = styled.div`
  width: 320px;
  display: flex;
  justify-content: center;
  align-items: center;
  border-top: 1px solid var(--base-color-border);
  margin-top: 20px;
  margin-bottom: 16px;
  padding-top: 16px;

  > p {
    font-size: 14px;
    line-height: 18px;
    color: var(--base-color-warm-grey);
  }

  p > a {
    color: var(--base-color-link);
    text-decoration: none;
  }

  @media (max-width: 600px) {
    width: 312px;
    margin-top: 81px;
    padding-top: 15.8px;
  }
`;

const RegisterFormWrapper = styled.div`
  width: 320px;
  height: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-top: 36px;
  margin-bottom: 40px;

  @media (max-width: 600px) {
    width: 312px;
    margin-top: 148px;
    margin-bottom: 32px;
  }
`;

const Form = styled.form`
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  padding-top: 23px;
`;

export default RegisterFormContainer;
