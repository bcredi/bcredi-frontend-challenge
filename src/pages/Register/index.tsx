import React, { useEffect, useState } from "react";
import Hero from "../../components/Hero";
import RegisterForm from "../../components/RegisterForm";
import { ReactComponent as Logo } from "../../assets/logo.svg";
import {
  HeroQuote,
  HeroAuthorInfo,
  PageWrapper,
  RegisterWapper,
  SuccessBar,
  Header,
  FlexGroup,
} from "../../styles";
import TextField from "../../components/TextField";
import { User, InputError } from "./types";
import Checkbox from "../../components/CheckBox";

export const rejectedCpfs = [
  "00000000000",
  "11111111111",
  "22222222222",
  "33333333333",
  "44444444444",
  "55555555555",
  "66666666666",
  "77777777777",
  "88888888888",
  "99999999999",
];

export function validateEmailFormat(email: string) {
  if (email !== "") {
    const match = email.match(
      // eslint-disable-next-line no-useless-escape
      /^(([^<>()[\]\\.,+`^*!'¨|;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
    );
    const correctFormat = match !== null;
    if (correctFormat) {
      return 1;
    } else {
      return "incorrect format";
    }
  } else {
    return "empty email";
  }
}

export function validateCpfFormat(cpf: string) {
  if (cpf !== "") {
    const match = cpf.match(/\d\d\d[.]\d\d\d[.]\d\d\d[-]\d\d$/);
    const correctFormat = match !== null;
    if (correctFormat) {
      const stringOnlyNumbers = cpf.replace(/[.-]/g, "");
      if (rejectedCpfs.includes(stringOnlyNumbers)) {
        return "repetitive numbers";
      } else {
        const cpfNumbersArray = stringOnlyNumbers.split("").map((n) => +n);

        const sumNumberFirstDigit = cpfNumbersArray
          .slice(0, 9)
          .map((n, index) => n * (10 - index))
          .reduce((a, b) => a + b, 0);

        const sumNumberSecondDigit = cpfNumbersArray
          .slice(0, 10)
          .map((n, index) => n * (11 - index))
          .reduce((a, b) => a + b, 0);

        var restDigitOne = (sumNumberFirstDigit * 10) % 11;
        var restDigitTwo = (sumNumberSecondDigit * 10) % 11;

        if (restDigitOne === 10 || restDigitOne === 11) {
          restDigitOne = 0;
        }
        if (restDigitTwo === 10 || restDigitTwo === 11) {
          restDigitTwo = 0;
        }
        if (
          restDigitOne !== cpfNumbersArray[9] ||
          restDigitTwo !== cpfNumbersArray[10]
        ) {
          return "rest does not match the digits";
        } else {
          return 1;
        }
      }
    } else {
      return "incorrect format";
    }
  } else {
    return "empty cpf";
  }
}

export function validateDateFormat(date: string) {
  if (date !== "") {
    const match = date.match(
      /^([1-9]|0[1-9]|[12][0-9]|3[01])[/]([1-9]|0[1-9]|1[012])[/](19|20)\d\d$/
    );
    const correctFormat = match !== null;

    if (correctFormat) {
      const currentYear = new Date().getFullYear();
      const monthMaxDays = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
      const [day, month, year] = date.split("/").map((s) => +s); // Conversão para int

      if (year <= currentYear) {
        // months - fev
        if (month === 1 || month > 2) {
          if (day > monthMaxDays[month - 1]) {
            return "day outside the month limits";
          } else {
            return 1;
          }
        }

        // fev
        if (month === 2) {
          const leapYear =
            (!(year % 4) && year % 100) || (!(year % 400) as boolean);
          if (!leapYear && day >= 29) {
            return "day outside the limits of February - not a leap year";
          } else if (leapYear && day > 29) {
            return "day outside the limits of February - leap year";
          } else {
            return 1;
          }
        }
      } else {
        return "year bigger than the current one";
      }
    } else {
      return "incorrect format";
    }
  } else {
    return "empty birthDate";
  }
}

export function validatePasswordFormat(password: string) {
  if (password !== "") {
    if (password.length >= 8) {
      return 1;
    } else {
      return "incorrect format";
    }
  } else {
    return "empty password";
  }
}

const Register = () => {
  // States de controle
  const [registered, setRegistered] = useState(false);

  // Dados obtidos através dos inputs
  const [userData, setUserData] = useState<User>({} as User);
  const [formErrors, setFormErrors] = useState<InputError>({} as InputError);
  const [readCheck, setReadCheck] = useState(false);

  // Error flags
  const [emailError, setEmailError] = useState(false);
  const [cpfError, setCpfError] = useState(false);
  const [birthDateError, setBirthDateError] = useState(false);
  const [passwordError, setPasswordError] = useState(false);
  const [readCheckError, setReadCheckError] = useState(false);

  useEffect(() => {
    if (formErrors.emailError) {
      setEmailError(true);
    } else if (formErrors.cpfError) {
      setCpfError(true);
    } else if (formErrors.birthDateError) {
      setBirthDateError(true);
    } else if (formErrors.passwordError) {
      setPasswordError(true);
    } else if (formErrors.readCheckError) {
      setReadCheckError(true);
    }
  }, [formErrors]);

  const handleRegisterSuccess = () => {
    console.log(userData);
    setRegistered(true);
    setTimeout(() => setRegistered(false), 3000);
  };

  const handleChange = (Event: React.ChangeEvent<HTMLInputElement>) => {
    const { name, value } = Event.target;
    setUserData({ ...userData, [name]: value });
  };

  return (
    <PageWrapper>
      <Hero>
        <HeroQuote>
          Obtive crédito para capital de giro. O processo foi bem sucedido e
          tudo que foi abordado, foi cumprido.
        </HeroQuote>
        <HeroAuthorInfo>Camila Bragança,</HeroAuthorInfo>
        <HeroAuthorInfo>Sideral Tecnologia</HeroAuthorInfo>
      </Hero>
      <RegisterWapper>
        <SuccessBar className={registered ? "visible" : ""}>
          Parabéns! Cadastro realizado com sucesso!
        </SuccessBar>
        <Header>
          <Logo className="logo" />
        </Header>
        <RegisterForm
          handleSubmit={handleRegisterSuccess}
          formTitle="Criar meu cadastro"
          formDescription="Para acompanhar sua contratação de crédito você utilizará seu e-mail e
        CPF."
          footer={
            <p>
              Já fiz o meu cadastro. <a href="/login">Entrar agora.</a>
            </p>
          }
          setFormErrors={setFormErrors}
          formErrors={formErrors}
          readCheck={readCheck}
          data={userData}
          setData={setUserData}
        >
          <TextField
            setEmpty={setEmailError}
            handleChange={handleChange}
            error={emailError}
            value={userData.email}
            name={"email"}
            placeholder={"ana.maria@email.com"}
            validateValueFormat={validateEmailFormat}
          />
          <FlexGroup>
            <TextField
              setEmpty={setCpfError}
              error={cpfError}
              handleChange={handleChange}
              maxLength={14}
              value={userData.cpf}
              name={"cpf"}
              placeholder={"000.000.000-00"}
              validateValueFormat={validateCpfFormat}
            />
            <TextField
              setEmpty={setBirthDateError}
              error={birthDateError}
              handleChange={handleChange}
              maxLength={10}
              value={userData.birthDate}
              name={"birthDate"}
              placeholder={"DD/MM/AAAA"}
              validateValueFormat={validateDateFormat}
            />
          </FlexGroup>
          <TextField
            setEmpty={setPasswordError}
            handleChange={handleChange}
            error={passwordError}
            minLength={8}
            value={userData.password}
            name={"password"}
            placeholder={"Cadastre uma senha"}
            validateValueFormat={validatePasswordFormat}
          />
          <Checkbox
            setNotChecked={setReadCheckError}
            handleChange={setReadCheck}
            error={readCheckError}
            value={readCheck}
          />
        </RegisterForm>
      </RegisterWapper>
    </PageWrapper>
  );
};

export default Register;
