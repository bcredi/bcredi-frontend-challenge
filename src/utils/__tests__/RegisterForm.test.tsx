import React from "react";
import RegisterPage from "../../pages/Register";
import renderer from "react-test-renderer";
import RegisterForm from "../../components/RegisterForm";
import { InputError, User } from "../../pages/Register/types";
import { shallow } from "enzyme";

describe("<RegisterPage />", () => {
  let component;

  beforeEach(() => {
    component = shallow(<RegisterPage />);
  });

  test("It should mount", () => {
    expect(component.length).toBe(1);
  });

  test("Checks if the component is correctly rendered", () => {
    const component = renderer
      .create(
        <RegisterForm
          formTitle="Test"
          formDescription="Test description"
          handleSubmit={() => {}}
          data={{} as User}
          formErrors={{} as InputError}
          setData={() => {}}
          setFormErrors={() => {}}
          readCheck={false}
        />
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
  
  test("should render the form title", () => {
    const wrapper = shallow(
      <RegisterForm
        formTitle="Test"
        formDescription="Test description"
        handleSubmit={() => {}}
        data={{} as User}
        formErrors={{} as InputError}
        setData={() => {}}
        setFormErrors={() => {}}
        readCheck={false}
      />
    );
    const title = wrapper.find(".form-title");
    const description = wrapper.find(".form-description")
    expect(title.text()).toBe("Test");
    expect(description.text()).toBe("Test description");
  });
});
