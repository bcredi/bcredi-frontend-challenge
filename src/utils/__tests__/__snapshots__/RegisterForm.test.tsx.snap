// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterPage /> Checks if the component is correctly rendered 1`] = `
<div
  className="sc-iBPRYJ dhIWtH"
>
  <div
    className="sc-eCssSg kndXy form-title"
  >
    Test
  </div>
  <div
    className="sc-jSgupP gNUGqr form-description"
  >
    Test description
  </div>
  <form
    className="sc-fubCfw iNVubo"
    name="register"
    onSubmit={[Function]}
  >
    <button
      className="sc-dlfnbm cXRuMu"
      type="submit"
    >
      <svg
        className="sc-hKgILt gDcxHx"
      >
        lock-solid.svg
      </svg>
      <p>
        Cadastrar
      </p>
    </button>
  </form>
  <div
    className="sc-gKsewC jzxsty form-footer"
  />
</div>
`;

exports[`Checks if the component is correctly rendered 1`] = `
<div
  className="sc-iBPRYJ dhIWtH"
>
  <div
    className="sc-eCssSg kndXy form-title"
  >
    Test
  </div>
  <div
    className="sc-jSgupP gNUGqr form-description"
  >
    Test description
  </div>
  <form
    className="sc-fubCfw iNVubo"
    name="register"
    onSubmit={[Function]}
  >
    <button
      className="sc-dlfnbm cXRuMu"
      type="submit"
    >
      <svg
        className="sc-hKgILt gDcxHx"
      >
        lock-solid.svg
      </svg>
      <p>
        Cadastrar
      </p>
    </button>
  </form>
  <div
    className="sc-gKsewC jzxsty form-footer"
  />
</div>
`;
